### ISSUE TITLE
role_delegation module is causing automated tests to fail

### ISSUE THREAD

**Issue by CraigClark** [bcgov/MFIN-Data-Catalogue#169]
**OP timer**
https://openplus.monday.com/boards/4092908516/pulses/5158134800

---

## Describe the bug

The `role_delegation` module is causing automated tests to fail. This happens because it creates permissions which are assigned in config export. When functional tests run, these permissions do not yet exist at the time the permissions config is imported. Drupal no longer allows roles to have permissions that do not exist.

We use this module to control what roles data set administrator can assign. We tried a few work arounds, but they all failed. 

We have created a role that can assign other roles in Drupal's UI for permissions. This is a temporary solution as a person could assign themselves the admin role and get full admin access. This isn't an issue if it is a few trusted employees, however if data_catalog_admin roles are broadly assigned, it will be a problem.

We have part of a solution, but it gets a bit complicated as configuring the roles a user can assign happens in code. We could build a UI for that, but it would fall outside Drupal's permissions page.

We are going to submit what we have to drupal.org in the hopes someone can get this fixed. In the mean time, we will keep this issue open.

**Comment by lkmorlan** [comment]
The bug is reported here:
https://www.drupal.org/project/role_delegation/issues/3354012

A work-around is to use the patch in this issue:
https://www.drupal.org/project/role_delegation/issues/3387139
This patch creates a hook that allows us to define the assignable roles. This would allow putting the roles in code:

```
/**
 * Implements hook_role_delegation_assignable_roles_alter().
 */
function bc_dc_role_delegation_assignable_roles_alter(array &$assignable_roles, array $all_roles, AccountInterface $account): void {
  if (!$account->hasPermission('assign bc_dc roles')) {
    return;
  }

  $roles_to_assign = [
    'data_catalogue_administrator',
    'data_catalogue_editor',
    'data_catalogue_user',
  ];
  foreach ($roles_to_assign as $role) {
    $assignable_roles[$role] = $all_roles[$role]->label();
  }
}
```

We could also use the hook and have a configuration UI to allow the roles to be stored config.

**Comment by CraigClark** [comment]
@lkmorlan I'm assigning to you and moving to ready for review as you mentioned in slack that you have a fix

**Comment by lkmorlan** [comment]
Fixed in #162 .
https://github.com/bcgov/MFIN-Data-Catalogue/commit/a49c98ba4b596b7b1eeb45d22284597373c73691

**Comment by CraigClark** [comment]
tested on dv21 and passed
